### PR TITLE
8293170: Improve encoding of the debuginfo nmethod section

### PR DESCRIPTION
**Pull request by bulasevich**
This is another pull request to replace https://github.com/openjdk/jdk/pull/10025 change which was blocked as not acceptable (see https://github.com/openjdk/jdk/pull/10025#pullrequestreview-1228216330)

The objections to change #10025 were:
- specialized algorithm for given data complicates things, makes it hard to learn, test and support
- algorithm is changed for DebugInfo, and the benefit is only for one type of data
- statistics of the debug info data can (will) change, breaking the optimization

The suggestion was:
- don't change the core algorithm, but add one on top or underneath the existing one, or reuse off-the-shelf zero-reduction schemes such as Cap'n Proto

With this change I propose a different approach. Instead of bit coding, the sequence of zeros in a data stream is encoded with a special character that normally never appears in Unsinged5 encoding, followed by a byte containing a number of zeros. In this way the updated algorithm is a pure extension of the existing encoding algorithm: data encoded without the zero-reduction trick is unpacked in the same way as before.

Currently there are several datasets affected by this change: Dependencies info, OopMap info, LineNumber info, Debug info. Only Debug info has a large number of zeros and gets a significant benefit. I experimented with the Cap'n Proto and lz4 algorithms on DebugInfo. The Unsinged5 algorithm has a better compression rate than these.

DebugInfo data size is reduced by ~20% (actually, 10-30%, depending on the application). Total nmethod size reduction is ~3%.

Performance impact: Renaisance and DaCapo benchmarks do not show any difference.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293170](https://bugs.openjdk.org/browse/JDK-8293170): Improve encoding of the debuginfo nmethod section (**Enhancement** - P3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12387/head:pull/12387` \
`$ git checkout pull/12387`

Update a local copy of the PR: \
`$ git checkout pull/12387` \
`$ git pull https://git.openjdk.org/jdk.git pull/12387/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12387`

View PR using the GUI difftool: \
`$ git pr show -t 12387`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12387.diff">https://git.openjdk.org/jdk/pull/12387.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12387#issuecomment-1544079749)

### CODE REVIEW

**Comment by bulasevich** [comment]
@rose00 @vnkozlov I would appreciate it if you could find time to review this change. Thank you very much.

**Comment by bulasevich** [comment]
kind reminder

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedReadStream.java [hunk 1]
```diff
@@ -87,6 +101,12 @@ public int readInt() {
                                     CompressedReadStream::setPosition);
   }
 
+  @Override
+  public void setPosition(int position) {
+    this.position = position;
```
Comment (plummercj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedReadStream.java:106]: Maybe a call to `super()` would be better.

File: src/hotspot/share/code/compressedStream.hpp [hunk 2]
```diff
@@ -81,20 +102,38 @@ class CompressedWriteStream : public CompressedStream {
...
     if (full()) grow();
     store(b);
   }
   void grow();
 
+  // Handle encoding of subsequent zeroes. Return true if
+  // input value is completely handled and no unsigned5 encoding required
+  bool handle_zero(juint value) {
+    if (value == 0) {
+      _zero_count = (_zero_count == 0xFF) ? 0 : _zero_count;
```
Comment (sspitsyn) [inline src/hotspot/share/code/compressedStream.hpp:118]: The case of `_zero_count` overflow is not clear. Apparently, I'm missing something here.
Current code is just clearing the previously counted `_zero_count`.
I'd expect some action like storing the current number of zeros or advancing the `_position`.
Do you have a test for this?

Comment (bulasevich) [reply src/hotspot/share/code/compressedStream.hpp:118]: Good question. It works like this. On each write(), values are stored in the array, either normally or as a chain optimization (two bytes: zero byte and a number of zeros). When we get another zero value, the chain is updated. When the chain is full, we exit this function and the value is written by UNSIGNED5::write_uint_grow().

To make things clear, I have updated this code:
```diff
-      _zero_count = (_zero_count == 0xFF) ? 0 : _zero_count;
+      if (_zero_count == 0xFF) { // biggest zero chain length is 255
+        _zero_count = 1;
+        // for now, write it as an ordinary value (UNSINGED5 encodes zero int as a single byte)
+        // the new zero sequence is started if there are more than two zero values in a raw
+        return false;
+      }
```

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedReadStream.java (hunk 1 above)
Comment (bulasevich) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedReadStream.java]: ok. thanks!

**Pull request closed without merging**